skills_ml/algorithms/embedding/train.py

```diff
@@ -55,83 +57,110 @@ def __init__(
+        tic = time()
+        try:
+            self.corpus_metadata = corpus_generator.metadata
+        except AttributeError:
+            self.corpus_metadata = None
+
+        if any([model.wv.vocab for model in self._models]):
+            logging.info("Model has been trained")
+            return 0
+
+        for model in self._models:
+            model.model_name = "_".join([
+                model.model_type,
+                self._model_hash(model, self.training_time, self.corpus_metadata)]
+            ) + '.model'
+
+        if self.model_type <= set(['word2vec', 'fasttext']):
+            batch_gen = BatchGenerator(corpus_generator, self.batch_size)
+            if n_processes == 1:
+                for i, batch in enumerate(batch_gen):
+                    logging.info("Training batch #{} ".format(i))
+                    self._models = [self._train_one_batch(model, batch) for model in self._models]
+            else:
+                with mp.Pool(processes=n_processes, maxtasksperchild=1) as pool:
+                    for i, batch in enumerate(batch_gen):
+                        logging.info("Training batch #{} ".format(i))
+                        partial_train = partial(self._train_one_batch, batch=batch, *args, **kwargs)
+                        self._models = pool.map(partial_train, self._models)
+
+        elif set(self.model_type) == set(['doc2vec']):
+            logging.info(f"Training {self.model_type}")
+            reiter_corpus_gen = Reiterable(corpus_generator)
+            self._models = [self._train_one_batch(model, reiter_corpus_gen) for model in self._models]
             if lookup:
-                self.lookup_dict = corpus_gen.lookup
-                self._model.lookup_dict = self.lookup_dict
+                self.lookup_dict = corpus_generator.lookup
+                for model in self._models:
+                    model.lookup_dict = self.lookup_dict
+        else:
+            raise TypeError("Doc2Vec model can only be trained alone with its own kind, not supporting training with other models")
 
-        self._model.model_name = self.model_name
+        logging.info(f"{', '.join([m.model_name for m in self._models])} are trained in {str(timedelta(seconds=time()-tic))}")
+
+    def _model_hash(self, model, training_time, corpus_metadata):
+        unique = {
+            "model_metadata": model.metadata,
+            "training_time": training_time,
+            "corpus_metadata": corpus_metadata
+        }
+        return filename_friendly_hash(unique)
 
     def save_model(self, storage=None):
         if storage is None:
             if self.model_storage is None:
                 raise AttributeError(f"'self.model_storage' should not be None if you want to save the model")
             ms = self.model_storage
-            ms.save_model(self._model, self.model_name)
+            for model in self._models:
```

Comment:
Unless I'm missing something, we cold reduce the branching by dong something like:

```
if storage:
   ms = ModelStorage(storage)
else:
   ms = self.model_storage

for model in self._models:
   ms.save_model
```